File: Pinball App Android/app/src/main/java/com/pillyliu/pinballandroid/practice/PracticeInsightsSection.kt
package com.pillyliu.pinballandroid.practice

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.getValue
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.pillyliu.pinballandroid.data.redactPlayerNameForDisplay
import com.pillyliu.pinballandroid.ui.CardContainer

@Composable
internal fun PracticeInsightsSection(
    store: PracticeStore,
    selectedGameSlug: String?,
    onSelectGameSlug: (String) -> Unit,
    insightsOpponentName: String,
    insightsOpponentOptions: List<String>,
    onInsightsOpponentNameChange: (String) -> Unit,
    headToHead: HeadToHeadComparison?,
    isLoadingHeadToHead: Boolean,
    onRefreshHeadToHead: () -> Unit,
) {
    val orderedGames = orderedGamesForDropdown(store.games, collapseByPracticeIdentity = true)
    val game = selectedGameSlug?.let { slug ->
        findGameByPracticeLookupKey(orderedGames, slug)
    } ?: orderedGames.firstOrNull()?.also { onSelectGameSlug(it.practiceKey) }

    if (game == null) {
        Text("No game data.")
        return
    }

    val availableSources = store.librarySources
    var gamePickerExpanded by remember { mutableStateOf(false) }
    Box {
        OutlinedButton(
            onClick = { gamePickerExpanded = true },
            modifier = Modifier.fillMaxWidth(),
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                Text(
                    text = store.gameName(game.practiceKey),
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                    modifier = Modifier.weight(1f),
                )
                Text("▼", style = MaterialTheme.typography.labelSmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
            }
        }
        DropdownMenu(
            expanded = gamePickerExpanded,
            onDismissRequest = { gamePickerExpanded = false },
        ) {
            if (availableSources.size > 1) {
                DropdownMenuItem(
                    text = {
                        Text(
                            (if (store.defaultPracticeSourceId == null) "✓ " else "") + "All games",
                            maxLines = 1,
                            overflow = TextOverflow.Ellipsis,
                        )
                    },
                    onClick = {
                        store.setPreferredLibrarySource(null)
                    },
                )
                availableSources.forEach { source ->
                    DropdownMenuItem(
                        text = {
                            Text(
                                (if (source.id == store.defaultPracticeSourceId) "✓ " else "") + source.name,
                                maxLines = 1,
                                overflow = TextOverflow.Ellipsis,
                            )
                        },
                        onClick = {
                            store.setPreferredLibrarySource(source.id)
                        },
                    )
                }
                HorizontalDivider()
            }
            orderedGames.forEach { insightsGame ->
                DropdownMenuItem(
                    text = { Text(insightsGame.name, maxLines = 1, overflow = TextOverflow.Ellipsis) },
                    onClick = {
                        gamePickerExpanded = false
                        onSelectGameSlug(insightsGame.practiceKey)
                    },
                )
            }
        }
    }
    CardContainer {
        Text("Stats", fontWeight = FontWeight.SemiBold)
        val summary = store.scoreSummaryFor(game.practiceKey)
        val trendValues = store.scoreTrendValues(game.practiceKey)
        if (summary == null) {
            Text("Log scores to unlock trends and consistency analytics.")
        } else {
            StatRow("Average", formatScore(summary.mean))
            StatRow("Median", formatScore(summary.median))
            StatRow("Floor", formatScore(summary.low))
            Text(
                "IQR: ${formatScore(summary.p25)} to ${formatScore(summary.targetHigh)}",
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
            Text(
                "Mode: Shows raw calendar spacing between score entries.",
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
            ScoreTrendSparkline(values = trendValues)
            Row(horizontalArrangement = Arrangement.spacedBy(8.dp), modifier = Modifier.fillMaxWidth()) {
                val spreadRatio = if (summary.median <= 0) 1.0 else (summary.targetHigh - summary.targetFloor) / summary.median
                DashboardMetricPill(
                    label = "Consistency",
                    value = if (spreadRatio >= 0.6) "High Risk" else "Stable",
                    modifier = Modifier.weight(1f),
                )
                DashboardMetricPill(
                    label = "Floor",
                    value = formatScore(summary.low),
                    modifier = Modifier.weight(1f),
                )
                DashboardMetricPill(
                    label = "Median",
                    value = formatScore(summary.median),
                    modifier = Modifier.weight(1f),
                )
            }
        }
    }

    CardContainer {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text("Head-to-Head", fontWeight = FontWeight.SemiBold, modifier = Modifier.weight(1f))
            TextButton(
                onClick = onRefreshHeadToHead,
                enabled = !isLoadingHeadToHead,
            ) { Text("Refresh") }
        }

        InsightsMenuDropdown(
            selectedLabel = if (insightsOpponentName.isBlank()) {
                "Select player"
            } else {
                redactPlayerNameForDisplay(insightsOpponentName)
            },
            options = listOf("" to "Select player") + insightsOpponentOptions.map { it to redactPlayerNameForDisplay(it) },
            onSelect = { pair ->
                onInsightsOpponentNameChange(pair.first)
            },
        )

        when {
            isLoadingHeadToHead -> Text("Loading player comparison...", style = MaterialTheme.typography.bodySmall)
            insightsOpponentName.isBlank() -> Text("Select a player above to enable player-vs-player views.", style = MaterialTheme.typography.bodySmall)
            headToHead == null -> Text(
                "No shared machine history yet between ${if (store.playerName.isBlank()) "you" else redactPlayerNameForDisplay(store.playerName)} and ${redactPlayerNameForDisplay(insightsOpponentName)}.",
                style = MaterialTheme.typography.bodySmall,
            )

            else -> {
                val comparison = headToHead
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp), modifier = Modifier.fillMaxWidth()) {
                    DashboardMetricPill(
                        label = "Games",
                        value = "${comparison.totalGamesCompared}",
                        modifier = Modifier.weight(1f),
                    )
                    DashboardMetricPill(
                        label = "You Lead",
                        value = "${comparison.gamesYouLeadByMean}",
                        modifier = Modifier.weight(1f),
                    )
                    DashboardMetricPill(
                        label = "Avg Delta",
                        value = shortSignedDelta(comparison.averageMeanDelta),
                        modifier = Modifier.weight(1f),
                    )
                }
                comparison.games.take(8).forEach { entry ->
                    HeadToHeadGameRow(entry)
                }
                if (comparison.games.size > 8) {
                    Text("Showing top 8 by mean delta.", style = MaterialTheme.typography.labelSmall)
                }
                val chartGames = comparison.games.take(8)
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(1.dp)
                        .background(MaterialTheme.colorScheme.onSurface.copy(alpha = 0.22f)),
                )
                HeadToHeadDeltaBars(
                    chartGames,
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(headToHeadPlotHeight(chartGames.size)),
                )
            }
        }
    }
}

File: Pinball App 2/Pinball App 2/practice/PracticeInsightsSection.swift
import SwiftUI

struct PracticeInsightsSectionView: View {
    let games: [PinballGame]
    let librarySources: [PinballLibrarySource]
    let selectedLibrarySourceID: String?
    let onSelectLibrarySourceID: (String?) -> Void
    @Binding var selectedGameID: String
    let scoreSummaryForGame: (String) -> ScoreSummary?
    let scoreTrendValuesForGame: (String) -> [Double]

    let playerName: String
    @Binding var opponentName: String
    let opponentOptions: [String]
    let isLoadingHeadToHead: Bool
    let headToHead: HeadToHeadComparison?
    let redactName: (String) -> String
    let onRefreshHeadToHead: () async -> Void
    let onRefreshOpponentOptions: () async -> Void

    private var selectedGame: PinballGame? {
        games.first(where: { $0.canonicalPracticeKey == selectedGameID || $0.id == selectedGameID })
    }

    private var selectedGameName: String {
        guard !selectedGameID.isEmpty else { return "Select game" }
        if let game = games.first(where: { $0.canonicalPracticeKey == selectedGameID || $0.id == selectedGameID }) {
            return game.name
        }
        return "Select game"
    }

    private var orderedInsightGames: [PinballGame] {
        orderedGamesForDropdown(games, collapseByPracticeIdentity: true)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 10) {
            insightsGameDropdown

            VStack(alignment: .leading, spacing: 8) {
                Text("Stats")
                    .font(.headline)

                if let gameID = selectedGame?.canonicalPracticeKey,
                   let summary = scoreSummaryForGame(gameID) {
                    HStack {
                        Text("Average")
                        Spacer()
                        Text(formattedScore(summary.average))
                            .fontWeight(.semibold)
                            .multilineTextAlignment(.trailing)
                    }
                    HStack {
                        Text("Median")
                        Spacer()
                        Text(formattedScore(summary.median))
                            .fontWeight(.semibold)
                            .multilineTextAlignment(.trailing)
                    }
                    HStack {
                        Text("Floor")
                        Spacer()
                        Text(formattedScore(summary.floor))
                            .fontWeight(.semibold)
                            .multilineTextAlignment(.trailing)
                    }
                    Text("IQR: \(formattedScore(summary.p25)) to \(formattedScore(summary.p75))")
                        .font(.footnote)
                        .foregroundStyle(.secondary)

                    Text("Mode: Shows raw calendar spacing between score entries.")
                        .font(.caption)
                        .foregroundStyle(.secondary)

                    ScoreTrendSparkline(values: scoreTrendValuesForGame(gameID))
                        .frame(height: 180)

                    HStack(spacing: 8) {
                        let spreadRatio = summary.median > 0 ? (summary.p75 - summary.floor) / summary.median : 0
                        MetricPill(label: "Consistency", value: spreadRatio >= 0.6 ? "High Risk" : "Stable")
                        MetricPill(label: "Floor", value: formattedScore(summary.floor))
                        MetricPill(label: "Median", value: formattedScore(summary.median))
                    }
                } else {
                    Text("Log scores to unlock trends and consistency analytics.")
                        .font(.footnote)
                        .foregroundStyle(.secondary)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(12)
            .appPanelStyle()

            VStack(alignment: .leading, spacing: 8) {
                HStack {
                    Text("Head-to-Head")
                        .font(.headline)
                    Spacer()
                    Button {
                        Task { await onRefreshHeadToHead() }
                    } label: {
                        Image(systemName: "arrow.clockwise")
                    }
                    .buttonStyle(.plain)
                    .disabled(isLoadingHeadToHead)
                }

                insightsOpponentDropdown

                if isLoadingHeadToHead {
                    ProgressView("Loading player comparison...")
                        .font(.footnote)
                } else if opponentName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                    Text("Select a player above to enable player-vs-player views.")
                        .font(.footnote)
                        .foregroundStyle(.secondary)
                } else if let scoped = headToHead {
                    HStack(spacing: 8) {
                        MetricPill(label: "Games", value: "\(scoped.totalGamesCompared)")
                        MetricPill(label: "You Lead", value: "\(scoped.gamesYouLeadByMean)")
                        MetricPill(label: "Avg Delta", value: signedScore(scoped.averageMeanDelta))
                    }

                    ForEach(Array(scoped.games.prefix(8))) { game in
                        HeadToHeadGameRow(game: game)
                    }
                    if scoped.games.count > 8 {
                        Text("Showing top 8 by mean delta.")
                            .font(.caption)
                            .foregroundStyle(.secondary)
                    }

                    let chartGames = Array(scoped.games.prefix(8))
                    Rectangle()
                        .fill(Color.white.opacity(0.22))
                        .frame(height: 1)
                    HeadToHeadDeltaBars(games: chartGames)
                        .frame(height: headToHeadPlotHeight(for: chartGames.count))
                } else {
                    Text("No shared machine history yet between \(playerName.isEmpty ? "you" : redactName(playerName)) and \(redactName(opponentName)).")
                        .font(.footnote)
                        .foregroundStyle(.secondary)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(12)
            .appPanelStyle()
        }
        .task {
            await onRefreshOpponentOptions()
        }
        .task(id: "\(playerName)|\(opponentName)") {
            await onRefreshHeadToHead()
        }
    }

    private var insightsGameDropdown: some View {
        Menu {
            if librarySources.count > 1 {
                Button((selectedLibrarySourceID == nil ? "✓ " : "") + "All games") {
                    onSelectLibrarySourceID(nil)
                }
                ForEach(librarySources) { source in
                    Button((source.id == selectedLibrarySourceID ? "✓ " : "") + source.name) {
                        onSelectLibrarySourceID(source.id)
                    }
                }
                Divider()
            }

            if orderedInsightGames.isEmpty {
                Text("No game data")
            } else {
                ForEach(orderedInsightGames) { game in
                    Button(game.name) {
                        selectedGameID = game.canonicalPracticeKey
                    }
                }
            }
        } label: {
            HStack(spacing: 8) {
                Text(selectedGameName)
                    .lineLimit(1)
                    .truncationMode(.tail)
                Spacer(minLength: 0)
                Image(systemName: "chevron.down")
                    .font(.caption.weight(.semibold))
                    .foregroundStyle(.secondary)
            }
            .padding(.horizontal, 10)
            .padding(.vertical, 8)
            .appControlStyle()
        }
    }

    private var insightsOpponentDropdown: some View {
        Menu {
            Button("Select player") {
                opponentName = ""
            }
            ForEach(opponentOptions, id: \.self) { name in
                Button(redactName(name)) {
                    opponentName = name
                }
            }
        } label: {
            HStack(spacing: 8) {
                Text(opponentName.isEmpty ? "Select player" : redactName(opponentName))
                    .lineLimit(1)
                    .truncationMode(.tail)
                Spacer(minLength: 0)
                Image(systemName: "chevron.down")
                    .font(.caption.weight(.semibold))
                    .foregroundStyle(.secondary)
            }
            .padding(.horizontal, 10)
            .padding(.vertical, 8)
            .appControlStyle()
        }
    }

    private func headToHeadPlotHeight(for count: Int) -> CGFloat {
        guard count > 0 else { return 170 }
        let rowHeight: CGFloat = 20
        let rowSpacing: CGFloat = 6
        let rows = CGFloat(count)
        let content = (rows * rowHeight) + (max(0, rows - 1) * rowSpacing) + 14
        return max(170, content)
    }

    private func formattedScore(_ value: Double) -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.maximumFractionDigits = 0
        return formatter.string(from: NSNumber(value: value)) ?? String(Int(value))
    }

    private func signedScore(_ value: Double) -> String {
        let sign = value > 0 ? "+" : ""
        return "\(sign)\(formattedScore(value))"
    }
}
